### PR TITLE
feat: `AbstractProgramContext` interface

### PR DESCRIPTION
**Pull request by Sai-prakash15**
*Issue #, if available:*

*Description of changes:*
Abstract the program context to define the functions which are necessary to build a circuit.
Add properties: probability, probabilities and gamma to noise operators.

*Testing done:*
Add unit test and Tested against all the existing test cases
## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/amazon-braket-default-simulator-python/blob/main/CONTRIBUTING.md) doc
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/amazon-braket-default-simulator-python/blob/main/CONTRIBUTING.md#commit-your-change)
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/amazon-braket-default-simulator-python/blob/main/README.md) and [API docs](https://github.com/aws/amazon-braket-default-simulator-python/blob/main/CONTRIBUTING.md#documentation-guidelines) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/braket/default_simulator/openqasm/parser/braket_pragmas.py [hunk 1]
```diff
@@ -213,7 +213,7 @@ def visitProbabilities(self, ctx: BraketPragmasParser.ProbabilitiesContext):
         return [float(prob.symbol.text) for prob in ctx.children[::2]]
 
 
-def parse_braket_pragma(pragma_body: str, qubit_table: "QubitTable"):
+def parse_braket_pragma(pragma_body: str, pragma_node_visitor=None):
```
Comment (ajberdy) [inline src/braket/default_simulator/openqasm/parser/braket_pragmas.py:216]: What's the motivation for this change? I don't think program context should need to import and instantiate a pragma node visitor, rather than just passing the qubit table

Comment (Sai-prakash15) [reply src/braket/default_simulator/openqasm/parser/braket_pragmas.py:216]: When I made the pragma node visitor abstract I changed it to accept pragma node visitor.
I felt like maybe in the future if someone wants to modify the behavior like this.

I will change it back to accept qubit table. I don't think anyone would change the behavior of pragma node visitor from program context

File: src/braket/default_simulator/openqasm/interpreter.py [hunk 2]
```diff
@@ -608,7 +608,7 @@ def handle_builtin_gate(
         qubits: List[Union[Identifier, IndexedIdentifier]],
         modifiers: List[QuantumGateModifier],
     ) -> None:
-        """Add unitary operation to the circuit"""
+        """Add unitary operation to the program"""
```
Comment (ajberdy) [inline src/braket/default_simulator/openqasm/interpreter.py:611]: I don't think I agree with this change. Here "Program" refers to the object being interpreted and "Circuit" the object being constructed; the operation is being added to the latter

File: src/braket/default_simulator/openqasm/program_context.py [hunk 3]
```diff
@@ -375,7 +377,7 @@ class ProgramContext:
 
     """
 
-    def __init__(self):
+    def __init__(self, program=Circuit()):
```
Comment (ajberdy) [inline src/braket/default_simulator/openqasm/program_context.py:380]: Similar to above, why not name this `circuit`?

File: src/braket/default_simulator/openqasm/parser/generated/BraketPragmasParser.py
Comment (ajberdy) [inline src/braket/default_simulator/openqasm/parser/generated/BraketPragmasParser.py:1]: We probably shouldn't be updating generated files

File: src/braket/default_simulator/simulator.py [hunk 4]
```diff
@@ -345,7 +346,7 @@ def run_openqasm(
                 are requested when shots>0.
         """
         is_file = openqasm_ir.source.endswith(".qasm")
-        interpreter = Interpreter()
+        interpreter = Interpreter(context=ProgramContext())
```
Comment (ajberdy) [inline src/braket/default_simulator/simulator.py:349]: This shouldn't be necessary, right? I'd like `Interpreter()` to work as expected with default arg values

File: src/braket/default_simulator/openqasm/interpreter.py (hunk 2 above)
Comment (Sai-prakash15) [reply src/braket/default_simulator/openqasm/interpreter.py:611]: I was thinking the same but Kshitij told me it could also support pulse programs in the future. And I did not know much about pulse programs. Hence changed the name to program

https://github.com/aws/amazon-braket-default-simulator-python/pull/162

Should I just change it back to circuit or leave it ?

Comment (ajberdy) [reply src/braket/default_simulator/openqasm/interpreter.py:611]: > OpenQASM encapsulates more information than what is described by the circuit class today. So it might be helpful to call this argument Program and pass in a circuit (which is a subset of a program)

The "circuit" here refers to what is built by the interpreter, which with today's architecture, is a level of functionality that we've been calling a circuit. Updates to the simulator to support more of a "program-level" scope are likely to involve a different architecture (ie. not building a single interpreted object). For example the [prototype program simulator](https://github.com/aws/amazon-braket-default-simulator-python/commit/38853b3edac022c9e16c7b7b04b7880ad9ec377a#diff-2e92ff2ca1084c457cd127f2a76a969d0e404853f4f5d74db8a0c8245267ce16R400) doesn't utilize the "circuit-building" part of program context, except for temporarily storing a circuit-level instruction buffer.

Given that context, I think `circuit` makes more sense. Feel free to weight in here @kshitijc

File: src/braket/default_simulator/openqasm/program_context.py [hunk 5]
```diff
@@ -571,6 +579,44 @@ def add_builtin_gate(
...
                 power *= mod.argument.value
+        self.add_gate_instruction(
+            gate_name, target, params, ctrl_modifiers=ctrl_modifiers, power=power
+        )
+
+    @abstractmethod
+    def add_gate_instruction(
+        self, gate_name: str, target: Tuple[int], params, ctrl_modifiers: List[int], power: int
+    ):
+        """Add gate instruction to the circuit"""
```
Comment (speller26) [inline src/braket/default_simulator/openqasm/program_context.py:590]: Also add argument information:
```
"""
Add gate instruction to the circuit.

Args:
    gate_name (str): ...
    target (int): ...
"""
```
Do this for other methods too

File: src/braket/default_simulator/openqasm/program_context.py [hunk 6]
```diff
@@ -375,7 +376,7 @@ class ProgramContext:
 
     """
 
-    def __init__(self):
+    def __init__(self, circuit=Circuit()):
```
Comment (speller26) [inline src/braket/default_simulator/openqasm/program_context.py:379]: Rather than having this in the base constructor, leave it to the subclasses.
```suggestion
    def __init__(self):
```

File: src/braket/default_simulator/openqasm/program_context.py [hunk 7]
```diff
@@ -384,7 +385,7 @@ def __init__(self):
         self.scope_manager = ScopeManager(self)
         self.inputs = {}
         self.num_qubits = 0
-        self.circuit = Circuit()
+        self.circuit = circuit
```
Comment (speller26) [inline src/braket/default_simulator/openqasm/program_context.py:388]: Remove this altogether...
```suggestion
```

File: src/braket/default_simulator/openqasm/program_context.py [hunk 8]
```diff
@@ -571,6 +579,44 @@ def add_builtin_gate(
...
+        """Add a custom Unitary instruction to the circuit"""
+
+    @abstractmethod
+    def add_noise_instruction(self, *args, **kwargs):
+        """Add a noise instruction to the circuit"""
+
+
+class ProgramContext(AbstractProgramContext):
+    def __init__(self):
+        super().__init__(Circuit())
```
Comment (speller26) [inline src/braket/default_simulator/openqasm/program_context.py:607]: ... And set the circuit here
```suggestion
    def __init__(self):
        super().__init__()
        self._circuit = Circuit()
```

File: src/braket/default_simulator/openqasm/program_context.py (hunk 7 above)
Comment (ajberdy) [reply src/braket/default_simulator/openqasm/program_context.py:388]: I don't think I agree with this. We could make the default None, but caching the "starting circuit" to `self.circuit` could totally be base class behavior. This is more than just a type annotation, it allows users to start not from scratch

File: src/braket/default_simulator/openqasm/program_context.py (hunk 6 above)
Comment (ajberdy) [reply src/braket/default_simulator/openqasm/program_context.py:379]: Maybe this could be
```suggestion
    def __init__(self, circuit=None):
```

File: src/braket/default_simulator/openqasm/program_context.py (hunk 8 above)
Comment (ajberdy) [reply src/braket/default_simulator/openqasm/program_context.py:607]: This behavior is likely duplicated in all subclasses, so to me it would make sense to keep that assignment in the base class

File: src/braket/default_simulator/openqasm/program_context.py (hunk 7 above)
Comment (speller26) [reply src/braket/default_simulator/openqasm/program_context.py:388]: But rather than keeping it in the abstract class, how about letting subclasses decide that? It feels kinda weird having the parent class decide that all base classes will build up a circuit _object_.

Comment (ajberdy) [reply src/braket/default_simulator/openqasm/program_context.py:388]: Hm I suppose that's true, just all the ones so far haha

File: src/braket/default_simulator/openqasm/program_context.py (hunk 8 above)
Comment (speller26) [reply src/braket/default_simulator/openqasm/program_context.py:607]: Is that necessarily the case though? I can totally conceive of a situation where an implementation looks like
```python
class MyContext(AbstractProgramContext):
    def __init__(self, ...):
        self._a = ...
        self._b = ...

    def exec(self):
        self._a.do_something()
        self._b.do_something()
```

File: src/braket/default_simulator/openqasm/program_context.py (hunk 7 above)
Comment (ajberdy) [reply src/braket/default_simulator/openqasm/program_context.py:388]: Even the monte-carlo/mid-circuit measurement sim has a circuit object in the program context, but that's just a circuit-level buffer of instructions that gets popped every time a measurement happens

File: src/braket/default_simulator/openqasm/program_context.py (hunk 8 above)
Comment (ajberdy) [reply src/braket/default_simulator/openqasm/program_context.py:607]: The interpreter class does have a `build_circuit` function, which returns `context.circuit`, but other than that there is no direct coupling between the interpreter and circuit classes. If someone wanted to build an interpreter/simulator that didn't have a build_circuit method, then yup I agree

Comment (speller26) [reply src/braket/default_simulator/openqasm/program_context.py]: Ah, in that case, it does make sense then.

File: src/braket/default_simulator/openqasm/program_context.py (hunk 6 above)
Comment (speller26) [reply src/braket/default_simulator/openqasm/program_context.py]: Yep, don't want to have mutable defaults.

File: src/braket/default_simulator/openqasm/program_context.py (hunk 8 above)
Comment (ajberdy) [reply src/braket/default_simulator/openqasm/program_context.py]: Arguably, if we carry on this line of thought, an abstract interpreter class wouldn't have a build_circuit class. For example the mcm interpreter doesn't utilize that method at all, but has a [simulate](https://github.com/aws/amazon-braket-default-simulator-python/blob/mcm-sim/src/braket/default_simulator/openqasm/native_interpreter.py#L39) method instead

File: src/braket/default_simulator/openqasm/program_context.py [hunk 9]
```diff
@@ -375,7 +387,7 @@ class ProgramContext:
 
     """
 
-    def __init__(self):
+    def __init__(self, program=Circuit()):
```
[the remark this answers is not in the document]
Comment (Sai-prakash15) [reply src/braket/default_simulator/openqasm/program_context.py]: I did not know that this was the reason.

Thanks for sharing it!

File: src/braket/default_simulator/openqasm/program_context.py (hunk 8 above)
Comment (speller26) [reply src/braket/default_simulator/openqasm/program_context.py]: Looking at `Interpreter`, it seems pretty deeply linked to the default simulator's `Circuit` class; no need to do it in this PR, but it seems to make sense to the visitor functionality from the build/run functionality.

**Review by speller26 — APPROVED** [review]

File: src/braket/default_simulator/openqasm/program_context.py [hunk 10]
```diff
@@ -571,6 +745,68 @@ def add_builtin_gate(
...
+                                        of the noise operation. (Optional)
+                - `probabilities` (List[float]): The probabilities associated with each possible outcome
+                                                of the noise operation. (Optional)
+        """
+
+
+class ProgramContext(AbstractProgramContext):
+    def __init__(self):
+        super().__init__()
+        self.circuit = Circuit()
```
Comment (ajberdy) [inline src/braket/default_simulator/openqasm/program_context.py:797]: if this specific program context class has a `self.circuit` attribute, its init should also accept a circuit argument of the same type

Comment (speller26) [reply src/braket/default_simulator/openqasm/program_context.py:797]: Quick fix: #183
